File: app/src/main/java/com/tetsworks/pocketlive/viewmodel/SequencerViewModel.java
package com.tetsworks.pocketlive.viewmodel;

import android.app.Application;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import com.tetsworks.pocketlive.audio.AudioEngine;
import com.tetsworks.pocketlive.audio.Sequencer;
import com.tetsworks.pocketlive.audio.Synthesizer;
import com.tetsworks.pocketlive.model.Track;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SequencerViewModel extends AndroidViewModel {

    public final AudioEngine audioEngine;
    public final Sequencer   sequencer;

    public final MutableLiveData<List<Track>> tracks      = new MutableLiveData<>();
    public final MutableLiveData<Integer>     currentStep = new MutableLiveData<>(-1);
    public final MutableLiveData<Boolean>     isPlaying   = new MutableLiveData<>(false);
    public final MutableLiveData<Integer>     bpm         = new MutableLiveData<>(120);

    private static final int[] COLORS = {
        0xFFE53935, 0xFF8E24AA, 0xFF1E88E5, 0xFF00ACC1,
        0xFF43A047, 0xFFFB8C00, 0xFFFFB300, 0xFF6D4C41
    };

    private static final int[] DEFAULT_INSTRUMENTS = {
        Synthesizer.KICK,
        Synthesizer.SNARE,
        Synthesizer.HIHAT,
        Synthesizer.CLAP
    };

    public SequencerViewModel(@NonNull Application application) {
        super(application);
        audioEngine = new AudioEngine(application);
        sequencer   = new Sequencer(audioEngine);

        List<Track> initial = new ArrayList<>();
        for (int i = 0; i < DEFAULT_INSTRUMENTS.length; i++) {
            int type = DEFAULT_INSTRUMENTS[i];
            String name = Synthesizer.EMOJIS[type] + " " + Synthesizer.NAMES[type];
            initial.add(new Track(i, name, type, COLORS[i]));
        }
        tracks.setValue(initial);
        sequencer.setTracks(initial);
        sequencer.setOnStepChangedListener(step -> currentStep.setValue(step));
    }

    private int nextId() {
        List<Track> list = tracks.getValue();
        if (list == null || list.isEmpty()) return 0;
        int max = 0;
        for (Track t : list) if (t.getId() > max) max = t.getId();
        return max + 1;
    }

    private int colorFor(int id) { return COLORS[id % COLORS.length]; }

    public void toggleStep(int trackId, int step) {
        List<Track> list = tracks.getValue();
        if (list == null) return;
        for (Track t : list) {
            if (t.getId() == trackId) {
                t.getSteps()[step] = !t.getSteps()[step];
                break;
            }
        }
        tracks.setValue(list);
    }

    public void toggleMute(int trackId) {
        List<Track> list = tracks.getValue();
        if (list == null) return;
        for (Track t : list) {
            if (t.getId() == trackId) { t.setMuted(!t.isMuted()); break; }
        }
        tracks.setValue(list);
    }

    public void setVolume(int trackId, float volume) {
        List<Track> list = tracks.getValue();
        if (list == null) return;
        for (Track t : list) {
            if (t.getId() == trackId) {
                t.setVolume(volume);
                audioEngine.updateVolume(t);
                break;
            }
        }
    }

    public void setBpm(int value) {
        sequencer.setBpm(value);
        bpm.setValue(sequencer.getBpm());
    }

    public void play()  { sequencer.play();  isPlaying.setValue(true);  }
    public void pause() { sequencer.pause(); isPlaying.setValue(false); }
    public void stop()  { sequencer.stop();  isPlaying.setValue(false); }

    public void addSynthTrack(int instrumentType) {
        List<Track> list = new ArrayList<>(tracks.getValue() != null ? tracks.getValue() : new ArrayList<>());
        int id = nextId();
        String name = Synthesizer.EMOJIS[instrumentType] + " " + Synthesizer.NAMES[instrumentType];
        list.add(new Track(id, name, instrumentType, colorFor(id)));
        tracks.setValue(list);
        sequencer.setTracks(list);
    }

    public void addCustomTrack(String name, Uri uri) {
        List<Track> list = new ArrayList<>(tracks.getValue() != null ? tracks.getValue() : new ArrayList<>());
        int id = nextId();
        Track track = new Track(id, "🎵 " + name, uri, colorFor(id));
        audioEngine.loadCustomAudio(id, uri, 1.0f);
        list.add(track);
        tracks.setValue(list);
        sequencer.setTracks(list);
    }

    public void removeTrack(int trackId) {
        List<Track> list = new ArrayList<>(tracks.getValue() != null ? tracks.getValue() : new ArrayList<>());
        list.removeIf(t -> t.getId() == trackId);
        tracks.setValue(list);
        sequencer.setTracks(list);
    }

    public void clearTrack(int trackId) {
        List<Track> list = tracks.getValue();
        if (list == null) return;
        for (Track t : list) {
            if (t.getId() == trackId) { Arrays.fill(t.getSteps(), false); break; }
        }
        tracks.setValue(list);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        sequencer.stop();
        audioEngine.release();
    }
}

File: app/src/main/java/com/tetsworks/pocketlive/viewmodel/SequencerViewModel.kt
package com.tetsworks.pocketlive.viewmodel

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.tetsworks.pocketlive.audio.AudioEngine
import com.tetsworks.pocketlive.audio.Sequencer
import com.tetsworks.pocketlive.audio.Synthesizer
import com.tetsworks.pocketlive.model.Track

class SequencerViewModel(app: Application) : AndroidViewModel(app) {

    val audioEngine = AudioEngine(app)
    val sequencer   = Sequencer(audioEngine)

    val tracks      = MutableLiveData<MutableList<Track>>()
    val currentStep = MutableLiveData<Int>(-1)
    val isPlaying   = MutableLiveData<Boolean>(false)
    val bpm         = MutableLiveData<Int>(120)

    private val trackColors = listOf(
        0xFFE53935.toInt(), 0xFF8E24AA.toInt(), 0xFF1E88E5.toInt(),
        0xFF00ACC1.toInt(), 0xFF43A047.toInt(), 0xFFFB8C00.toInt(),
        0xFFFFB300.toInt(), 0xFF6D4C41.toInt()
    )

    // Instrumentos padrão do sintetizador
    private val defaultInstruments = listOf(
        Synthesizer.InstrumentType.KICK,
        Synthesizer.InstrumentType.SNARE,
        Synthesizer.InstrumentType.HIHAT_CLOSED,
        Synthesizer.InstrumentType.CLAP,
    )

    init {
        val initialTracks = defaultInstruments.mapIndexed { i, type ->
            Track(
                id = i,
                name = "${type.emoji} ${type.displayName}",
                instrumentType = type,
                color = trackColors[i % trackColors.size]
            )
        }.toMutableList()

        tracks.value = initialTracks
        sequencer.setTracks(initialTracks)
        sequencer.onStepChanged = { step -> currentStep.value = step }
    }

    fun toggleStep(trackId: Int, step: Int) {
        tracks.value?.let { list ->
            list.find { it.id == trackId }?.steps?.let { steps ->
                steps[step] = !steps[step]
                tracks.value = list
            }
        }
    }

    fun toggleMute(trackId: Int) {
        tracks.value?.let { list ->
            list.find { it.id == trackId }?.let { it.isMuted = !it.isMuted }
            tracks.value = list
        }
    }

    fun setVolume(trackId: Int, volume: Float) {
        tracks.value?.let { list ->
            list.find { it.id == trackId }?.let {
                it.volume = volume
                audioEngine.updateVolume(it)
            }
        }
    }

    fun setPitch(trackId: Int, pitch: Float) {
        tracks.value?.let { list ->
            list.find { it.id == trackId }?.pitch = pitch
            tracks.value = list
        }
    }

    fun setBpm(value: Int) {
        sequencer.bpm = value
        bpm.value = sequencer.bpm
    }

    fun play()  { sequencer.play();  isPlaying.value = true  }
    fun pause() { sequencer.pause(); isPlaying.value = false }
    fun stop()  { sequencer.stop();  isPlaying.value = false }

    /** Adiciona instrumento sintetizado */
    fun addSynthTrack(type: Synthesizer.InstrumentType) {
        val list  = tracks.value ?: mutableListOf()
        val newId = (list.maxOfOrNull { it.id } ?: -1) + 1
        val track = Track(
            id = newId,
            name = "${type.emoji} ${type.displayName}",
            instrumentType = type,
            color = trackColors[newId % trackColors.size]
        )
        list.add(track)
        tracks.value = list
        sequencer.setTracks(list)
    }

    /** Adiciona áudio customizado importado (Fase 2) */
    fun addCustomTrack(name: String, uri: Uri) {
        val list  = tracks.value ?: mutableListOf()
        val newId = (list.maxOfOrNull { it.id } ?: -1) + 1
        val track = Track(
            id = newId,
            name = "🎵 $name",
            customAudioUri = uri,
            color = trackColors[newId % trackColors.size]
        )
        audioEngine.loadCustomAudio(newId, uri)
        list.add(track)
        tracks.value = list
        sequencer.setTracks(list)
    }

    fun removeTrack(trackId: Int) {
        val list = tracks.value ?: return
        list.removeAll { it.id == trackId }
        tracks.value = list
        sequencer.setTracks(list)
    }

    fun clearTrack(trackId: Int) {
        tracks.value?.let { list ->
            list.find { it.id == trackId }?.steps = BooleanArray(16) { false }
            tracks.value = list
        }
    }

    override fun onCleared() {
        super.onCleared()
        sequencer.stop()
        audioEngine.release()
    }
}
